### ISSUE TITLE
Event Bus - Add Task support

### ISSUE THREAD

**Issue by orchardbot** [OrchardCMS/Orchard#4846]
@sfmskywalker created:
https://orchard.codeplex.com/workitem/21017

When I implement an event handler with a method that returns something other than void, invoking that event handlers fails with an InvalidOperationException: Sequence contains no elements error.

This is because the EventsInterceptor only returns the correct type if the return type is void.
Fixing this for all return types is a bit tricky.
We could do it naively and simply return the first result, but if the return type is an array of strings for example, this wouldn't work - somewhere in the chain the return type is cast to IEnumerable, and this somehow causes the returned string array to be flattened a level, so that now the first element is a string, instead of a string array.

Typically this is not an issue, since a common pattern is to just return void, and accept some context argument which is used as both input and output.

However, I have a case where I want my handler to return a Task, so that I can leverage the async/await keywords all the way up to my controller.

To support this, I propose we add a check to the EventsIntercepter.Adjust method.
Currently, that code looks like this:
public static object Adjust(IEnumerable results, Type returnType) {
    if (returnType == typeof(void) ||
        results == null ||
        results.GetType() == returnType) {
        return results;
    }

```
// acquire method:
// static IEnumerable<T> IEnumerable.OfType<T>(this IEnumerable source)
// where T is from returnType's IEnumerable<T>
var enumerableOfTypeT = _enumerableOfTypeTDictionary.GetOrAdd( returnType, type => typeof(Enumerable).GetGenericMethod("OfType", type.GetGenericArguments(), new[] { typeof(IEnumerable) }, typeof(IEnumerable<>)));
return enumerableOfTypeT.Invoke(null, new[] { results });
```

}

After my proposed change, it would look like this (look for my comments):
public static object Adjust(IEnumerable results, Type returnType) {
    if (returnType == typeof(void) ||
        returnType == typeof(Task) ||
        results == null ||
        results.GetType() == returnType) {
        return results;
    }

```
// Check if the return type is a Task.
if (returnType == typeof (Task)) {
    return results.First();
}

// acquire method:
// static IEnumerable<T> IEnumerable.OfType<T>(this IEnumerable source)
// where T is from returnType's IEnumerable<T>
var enumerableOfTypeT = _enumerableOfTypeTDictionary.GetOrAdd( returnType, type => typeof(Enumerable).GetGenericMethod("OfType", type.GetGenericArguments(), new[] { typeof(IEnumerable) }, typeof(IEnumerable<>)));
return enumerableOfTypeT.Invoke(null, new[] { results });
```

}

Although this change won't fix the issue for any other return type, it will at least fix it for Task, which I think is a valid use case.

I tested this and it works nicely.

**Comment by Piedone** [comment]
Why not fix the Task return? Is it because fire and forget async is generally considered OK for event handlers? @sebastienros @sfmskywalker

**Comment by sfmskywalker** [comment]
I don't remember why, but my use case was having the ability to fire multiple events in parallel, and await them before returning an action result. Fire and forget is not an option, since I'd want to inspect some context arguments that I passed to these handlers.

**Comment by Piedone** [comment]
Even if you don't want to execute events in parallel it would be better if they could be awaited: if you write an async void event handler method (that you have to do if you use async code inside) you'll have to handle the work context yourself which is messy.

**Comment by sfmskywalker** [comment]
Yeah exactly. Another reason to add Task support.

**Comment by pszmyd** [comment]
Is this still up? As far as I remember calling an event handler with a return value of type `T` will always return an `IEnumerable<T>` (as there can be multiple handlers returning results). If a `T` is already an `IEnumerable<T>`, then the final result will also be an `IEnumerable<T>` with all the results from all handlers combined.

**Comment by pszmyd** [comment]
So `Task` type should already be supported.

**Comment by sfmskywalker** [comment]
That's what I thought as well, but the exception mentioned in the original post was thrown.
